Makefile

```diff
@@ -76,62 +76,33 @@ ifeq ($(NO_DISCOVER),true)
 	BASE_CMD += --nodiscover
 endif
 
-# Build suburl strings for slice specific subclient groups
-# WARNING: Only connect to dom/sub clients over a trusted network.
-ifeq ($(REGION),2)
-	PRIME_SUBS += ,,ws://127.0.0.1:$(REGION_$(REGION)_PORT_WS)
-endif
-ifeq ($(REGION),1)
-	PRIME_SUBS += ,ws://127.0.0.1:$(REGION_$(REGION)_PORT_WS),
-endif
-ifeq ($(REGION),0)
-	PRIME_SUBS += ws://127.0.0.1:$(REGION_$(REGION)_PORT_WS),,
-endif
-ifeq ($(ZONE),2)
-	REGION_SUBS =,,ws://127.0.0.1:$(ZONE_$(REGION)_$(ZONE)_PORT_WS)
-endif
-ifeq ($(ZONE),1)
-	REGION_SUBS =,ws://127.0.0.1:$(ZONE_$(REGION)_$(ZONE)_PORT_WS),
-endif
-ifeq ($(ZONE),0)
-	REGION_SUBS =ws://127.0.0.1:$(ZONE_$(REGION)_$(ZONE)_PORT_WS),,
-endif
+build_sequence = $(if $(filter $1,$2),,$1 $(call build_sequence,$(shell expr $1 + 1),$2))
+
+# Generate the sequence outside of any target
+WIDTH_SEQ := $(call build_sequence,0,$(WIDTH))
+
+# Step 2: Execute the commands in the run target's recipe
+# Command for the prime node
+PRIME_CMD := nohup $(BASE_CMD) --port $(PRIME_PORT_TCP) --http.port $(PRIME_PORT_HTTP) --ws.port $(PRIME_PORT_WS) --sub.urls $(PRIME_SUB_URLS) >> nodelogs/prime.log 2>&1 &
+
+# Template for the region nodes
+define REGION_CMD
+nohup $(BASE_CMD) --port $(REGION_$(1)_PORT_TCP) --http.port $(REGION_$(1)_PORT_HTTP) --ws.port $(REGION_$(1)_PORT_WS) --dom.url $(REGION_$(1)_DOM_URL):$(PRIME_PORT_WS) --sub.urls $(REGION_$(1)_SUB_URLS) --region $(1) >> nodelogs/region-$(1).log 2>&1 &
+endef
 
-# Build specific prime, region, and zone commands for run-slice
-PRIME_CMD = $(BASE_CMD) --port $(PRIME_PORT_TCP)
-PRIME_CMD += --http.port $(PRIME_PORT_HTTP)
-PRIME_CMD += --ws.port $(PRIME_PORT_WS)
-PRIME_CMD += --sub.urls "$(PRIME_SUBS)"
-PRIME_LOG_FILE = nodelogs/prime.log
-REGION_CMD = $(BASE_CMD) --region $(REGION) --port $(REGION_$(REGION)_PORT_TCP)
-REGION_CMD += --http.port $(REGION_$(REGION)_PORT_HTTP)
-REGION_CMD += --ws.port $(REGION_$(REGION)_PORT_WS)
-REGION_CMD += --dom.url ws://127.0.0.1:$(PRIME_PORT_WS)
-REGION_CMD += --sub.urls "$(REGION_SUBS)"
-REGION_LOG_FILE = nodelogs/region-$(REGION).log
-ZONE_CMD = $(BASE_CMD) --region $(REGION) --zone $(ZONE) --miner.etherbase $(ZONE_$(REGION)_$(ZONE)_COINBASE) --port $(ZONE_$(REGION)_$(ZONE)_PORT_TCP)
-ZONE_CMD += --http.port $(ZONE_$(REGION)_$(ZONE)_PORT_HTTP)
-ZONE_CMD += --ws.port $(ZONE_$(REGION)_$(ZONE)_PORT_WS)
-ZONE_CMD += --dom.url ws://127.0.0.1:$(REGION_$(REGION)_PORT_WS)
-ZONE_LOG_FILE = nodelogs/zone-$(REGION)-$(ZONE).log
+# Template for the zone nodes
+define ZONE_CMD
+nohup $(BASE_CMD) --miner.etherbase $(ZONE_$(1)_$(2)_COINBASE) --port $(ZONE_$(1)_$(2)_PORT_TCP) --http.port $(ZONE_$(1)_$(2)_PORT_HTTP) --ws.port $(ZONE_$(1)_$(2)_PORT_WS) --dom.url $(ZONE_$(1)_$(2)_DOM_URL):$(REGION_$(1)_PORT_WS) --region $(1) --zone $(2) >> nodelogs/zone-$(1)-$(2).log 2>&1 &
```

Comment:
why are these numbers hard coded to 1,2 ?

Reply:
Or is it the function argument index ?

Reply:
Function argument index. ZONE_CMD takes in region, zone